Tuya low-energy wi-fi devices support

SaschaKP opened the pull request:
# What does this implement/fix?

this pull request implements tuya low-energy devices that uses wifi modules combined with an MCU (battery powered - aka version 0)
requirements in configuration for this to work: setup priority is essential,  below an example

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other


**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [x] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
preferences:
  flash_write_interval: 
    seconds: 3

safe_mode:
  disabled: True

esphome:
  name: heiman-smoke-ws2sa-5
  friendly_name: tuya smoke detector
  name_add_mac_suffix: false

external_components:
  - source: github://pr#8953
    components: [tuya]
    refresh: 1h

bk72xx:
  board: cbu

logger:
  level: WARN #the lower the better for speed sake
  baud_rate: 0

mqtt:
  username: !secret mqtt_user
  password: !secret mqtt_password
  broker: [redacted]
  #down below we keep the device always available, even in offline, the MCU series exampled updates every 24 hours
  will_message: 
    topic: 
    payload: online
  birth_message: 
    topic: 
    payload: online

ota:
  platform: esphome
  password: !secret ota

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

switch:
  - platform: tuya
    id: snooze
    entity_category: config
    device_class: switch
    icon: mdi:volume
    name: Snooze Alarm
    switch_datapoint: 16
    retain: True

binary_sensor:
  - platform: template
    device_class: smoke
    name: Smoke Alarm
    id: smokealarm
    retain: True

time:
  - platform: sntp
    timezone: Europe/Rome
    servers: [redacted]
    on_time_sync: 
      - lambda: |-          
          id(sync) = true;

text_sensor:
  - platform: template
    name: "Last Restart"
    id: last_restart_mqtt
    update_interval: never
    device_class: timestamp
    icon: "mdi:clock"
    entity_category: diagnostic

globals:
  - id: updated
    type: bool
    restore_value: False
    initial_value: "false"
  - id: sync
    type: bool
    restore_value: False
    initial_value: "false"

sensor:
  - platform: wifi_signal
    name: Signal strength
    id: sensor_rssi
    update_interval: 500ms
    entity_category: diagnostic

  - platform: uptime
    type: timestamp
    id: last_restart
    entity_category: diagnostic
    internal: True
    on_value:
      - lambda: |-
          if(id(updated) == false && id(sync) == true)
          {
            id(updated) = true;
            char result[32];
            time_t ts_uptime = id(last_restart).state;
            strftime(result, sizeof(result), "%Y-%m-%dT%H:%M:%S+00:00", gmtime(&ts_uptime));
            id(last_restart_mqtt).publish_state(result);
          }

  - platform: tuya
    sensor_datapoint: 1
    name: Smoke Data
    id: smokedata
    internal: True
    on_value: 
      then:
        - lambda: |-
            if(x > 0)
              id(smokealarm).publish_state(false);
            else
              id(smokealarm).publish_state(true);

  - platform: tuya
    sensor_datapoint: 9
    name: Autotest
    id: autotest
    internal: True
  
  - platform: tuya
    sensor_datapoint: 14
    name: Battery Status
    id: battstate
    internal: True
  
  - platform: tuya
    sensor_datapoint: 15
    name: Battery Level
    state_class: measurement
    device_class: battery
    entity_category: diagnostic
    unit_of_measurement: "%"
    accuracy_decimals: 0
    id: battlevel
    retain: True

uart:
  - id: tuya_uart
    rx_pin: RX1
    tx_pin: TX1
    baud_rate: 9600
    parity: NONE
    stop_bits: 1
    #this is ABSOLUTELY REQUIRED for it to work, uart and tuya have precedence, in 2 seconds without response the device will deactivate itself
    setup_priority: 790

tuya:
  #required for this kind of devices
  low_energy: true
  #this is ABSOLUTELY REQUIRED for it to work, uart and tuya have precedence, in 2 seconds without response the device will deactivate itself
  setup_priority: 780

#USING API mode is not suggested, since it takes too much time for the CBU
#using API on esp32 is okay, just avoid it and use mqtt if you're on a less performing chipset
#this template is for the Heiman smoke ws2sa-5 that is exactly equivalent to shelly plus smoke
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

SaschaKP committed:
this pull request implements tuya low-energy devices that uses wifi modules combined with an MCU (battery powered - aka version 0)
requirements in configuration for this to work is the flag low_energy: true and reduce to IO the setup-priority of UART and tuya components to IO level

SaschaKP committed:
implement WIFI_PAIR for low_energy, WIFI_RESET is for long button press, and should be as a return to factory mode/settings, but at the moment it's not implemented

SaschaKP committed:
fix for tests by CI

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
- in init skip first attempt to give the time for the device to start correctly and avoid sending null data to serial
- re-order names for ENUM
- set DATAPOINT_DELIVER for battery powered devices correctly

SaschaKP committed:
- space and similar corrections for CI checks

SaschaKP committed:
- fixes indentation, required

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
upon returning online, sometime the device asks for a wireless test, we need to answer that the test is successful and send the payload 0x01 (success) and signal strength in percentage (0x32 for 50%, at the moment it's not calculated), only then the device will answer to dpId upon receiving the "connected" to cloud wireless status package

SaschaKP committed:
Merge branch 'dev' of https://github.com/SaschaKP/esphome into dev

SaschaKP committed:
Merge branch 'dev' into dev

aselafernando commented:
Rather than a low_energy configuration option, it should use a version parameter as suggested here https://github.com/esphome/esphome/pull/7551

SaschaKP committed:
Merge branch 'dev' into dev

ssieb commented:
I think it needs to be a separate component.  The state machine for battery powered devices is very different from normal Tuya devices.  The interaction is more driven by the external MCU.

aselafernando committed:
Tuya V0 Initialized Callback

Added call to dump_config and initialized_callback_ upon completion of initialization when operating with version 0 low energy. Also changed reference to cut_cloud_ to this->cut_cloud_mode_ to follow existing coding style.

aselafernando commented:
@SaschaKP I tested the code with an ESP8266 and Tuya v0 MCU in a PIR sensor and it works well. I did make some changes which I have created a PR on your repo for https://github.com/SaschaKP/esphome/pull/1

I added call to dump_config and initialized_callback_ upon completion of initialization when operating in this mode. I also changed references to cut_cloud_mode_ to this->cut_cloud_mode_ to follow existing coding style.

YAML for the ESP8266, ensures log output is on the second UART which is TX only and UART0 is for the Tuya MCU:
```
esp8266:
  board: esp01_1m
  framework:
    version: recommended

logger:
  level: VERBOSE
  hardware_uart: UART1

uart:
  - tx_pin: GPIO01
    rx_pin: GPIO03
    baud_rate: 9600

tuya:
  low_energy: true
```

@ssieb Not sure it needs to be a separate component, given the protocol structure seems to be shared across versions, it's mainly the initialisation sequence that is different and the numbering of the commands i.e. Command Type 0x05 is DATAPOINT_REPORT_SYNC in V0 whereas it would the WIFI_SELECT normally. The same codebase seems to be able to be reused without issue.

SaschaKP committed:
Merge branch 'esphome:dev' into dev

SaschaKP committed:
Merge pull request #1 from aselafernando/patch-4

Tuya V0 Initialized Callback

SaschaKP committed:
Merge branch 'esphome:dev' into dev

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
Merge branch 'esphome:dev' into dev

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
postpone responses to allow other background things to complete before device is shutdown by MCU

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

SaschaKP committed:
Merge branch 'esphome:dev' into dev

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
report true wifi quality in percent to MCU on wifi quality requests, limiting from minimum 1% up to the maximum 100%

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
Merge branch 'dev' into dev

SaschaKP committed:
Merge branch 'esphome:dev' into dev
